Core/PARStore.m

```diff
@@ -2126,64 +2208,110 @@ - (NSNumber *)mostRecentTimestampForKey:(NSString *)key
-        {
-            logsRequest.predicate = [NSPredicate predicateWithFormat:@"%K > %@", TimestampAttributeName, timestampLimit];
-        }
-        logsRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:TimestampAttributeName ascending:YES]];
-        logsRequest.resultType = NSDictionaryResultType;
-        NSArray *logs = [moc executeFetchRequest:logsRequest error:&errorLogs];
-        if (!logs)
-        {
-            ErrorLog(@"Error fetching logs for store at path '%@' because of error: %@", [self.storeURL path], errorLogs);
-            return;
-        }
-        
-        // logs --> changes
-        for (NSDictionary *logDictionary in logs)
-        {
-            NSNumber *timestamp = logDictionary[TimestampAttributeName];
-            NSNumber *parentTimestamp = logDictionary[ParentTimestampAttributeName];
-            NSString *key = logDictionary[KeyAttributeName];
-            NSData *blob = logDictionary[BlobAttributeName];
-            id propertyList = [self propertyListFromData:blob error:NULL];
-            if (timestamp != nil && key != nil && blob != nil)
-            {
-                PARChange *change = [PARChange changeWithTimestamp:timestamp parentTimestamp:parentTimestamp key:key propertyList:propertyList];
-                [changes addObject:change];
-            }
-        }
-        
-        [self closeDatabaseSoon];
-    }];
+     {
+         NSManagedObjectContext *moc = [self managedObjectContext];
+         if (moc == nil)
+         {
+             return;
+         }
+         
+         // From the documentation for `includesPendingChanges`: "A value of YES is not supported in conjunction with the result type NSDictionaryResultType, including calculation of aggregate results (such as max and min). For dictionaries, the array returned from the fetch reflects the current state in the persistent store, and does not take into account any pending changes, insertions, or deletions in the context."
+         // this means we need to save pending changes first to make sure they show up in the query
+         [self _save:NULL];
+         
+         // fetch Log rows in timestamp order, starting at `timestampLimit`
+         NSError *errorLogs = nil;
+         NSFetchRequest *logsRequest = [NSFetchRequest fetchRequestWithEntityName:LogEntityName];
+         
+         // Determine affected stores, based on device identifiers
+         if (fetchDeviceIdentifier == nil) {
+             logsRequest.affectedStores = nil; // All stores
+         }
+         else if ([fetchDeviceIdentifier isEqualToString:self.deviceIdentifier])
+         {
+             logsRequest.affectedStores = @[self.readwriteDatabase]; // Local store
+         }
+         else {
+             // Filter stores to find one that matches the device.
+             NSArray *eligibleStores = [self.readonlyDatabases filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^(NSPersistentStore *store, NSDictionary *bindings) {
```

Comment:
Wow, looking at this line after having been in the swift world for a while really does make objc look archaic!